File: Algorithm/Python/096. Unique Binary Search Trees.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-06-05 11:37:16
# @Last modified by:   WuLC
# @Last Modified time: 2016-06-06 14:03:29
# @Email: liangchaowu5@gmail.com

# method 1,divide and conquer
class Solution(object):
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        result = 0
        if n<=1:
            return 1
        for i in xrange(n/2):
            left = self.numTrees(i)
            right = self.numTrees(n-i-1)
            result += left * right
        result *= 2
        if n%2 == 1:
            result += pow(self.numTrees(n/2),2)
        return result


# method 2,DP
class Solution(object):
    def numTrees(self, n):
        """
        :type n: int
        :rtype: int
        """
        dp = [0 for i in xrange(n+1)]
        dp[0],dp[1] = 1, 1
        for i in xrange(2,n+1):
            for j in xrange(i):
                dp[i] += dp[j]*dp[i-j-1]
        return dp[n]

File: Algorithm/Java/096. Unique Binary Search Trees.java
/*
* @Author: WuLC
* @Date:   2017-06-20 10:26:55
* @Last Modified by:   WuLC
* @Last Modified time: 2017-06-20 12:22:52
* @Email: liangchaowu5@gmail.com
*/

// recursive with cache
public class Solution 
{
    private Map<Integer, Integer> cache = new HashMap<Integer, Integer>(); // TLE without cache
    public int numTrees(int n) 
    {
        if (cache.containsKey(n)) return cache.get(n);
        if (n <= 1) return n;
        int count = 0, left, right;
        for (int i = 0; i < n; i++)
        {
            left = numTrees(i);
            right = numTrees(n-i-1);
            if (left != 0 && right != 0) count += left * right;
            else count += left+right;
        }
        cache.put(n, count);
        return count;
    }
}


// dp without recursive
public class Solution 
{
    // dp[i] is the number of unique BST for i nodes
    private List<Integer> dp = new ArrayList<Integer>(); 
    public int numTrees(int n) 
    {
        if (dp.size() == 0) 
        {
            dp.add(0);
            dp.add(1);
        }
        for (int i = dp.size(); i <= n; i++)
        {
            int count = 0;
            for (int j = 0; j < i; j++) // number of nodes on the left sub tree
            {
                if ( j == 0 || j == i - 1) count += dp.get(i - 1);
                else count += dp.get(j) * dp.get(i-j-1);
            }
            dp.add(count);
        }
        return dp.get(n);
    }
}